### PR TITLE
Refactor loser tree  code in SortPreservingMerge per PR comments

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

re https://github.com/apache/arrow-datafusion/issues/4300.

# Rationale for this change
1. I wanted to get the merge speed improvements into DataFusion
2. I wanted an excuse to work on the code myself for a bit

# What changes are included in this PR?

Implements suggestions from @tustvold  and @viirya  in https://github.com/apache/arrow-datafusion/pull/4301 

# Are these changes tested?

covered by existing tests

TODO: need to run benchmarks

# Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs [hunk 1]
```diff
@@ -660,6 +609,112 @@ impl SortPreservingMergeStream {
...
+    }
+}
+
+/// The result of updating the loser tree. It is the same as an Option
+/// but with specific names for easier readability
+enum TreeUpdate {
+    /// The tree update could not be completed (e.g. the input was not
+    /// ready or had an error). The caller should return the `Poll`
+    /// result to its caller
+    Incomplete(Poll<Option<ArrowResult<RecordBatch>>>),
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:715]:
```suggestion
    Pending,
    
    Error(ArrowError),
```

Given we never seem to return `TreeUpdate::Incomplete(Poll::Ready(None))` or `TreeUpdate::Incomplete(Poll::Ready(Some(Ok(_))))`

Comment (alamb) [reply datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs]: in 1bdb25af1

**Comment by alamb** [comment]
Here are performance results. It is somewhat of a mixed bag (some show a few percent less) results are 
[results.txt](https://github.com/apache/arrow-datafusion/files/10108103/results.txt)

The first run is the second time I ran `cargo bench` against 0d334cf7 (aka no code change) and it reports some regressions and then the second run is with the changes in this PR

**Review by alamb — COMMENTED** [review]
Given this PR makes the code more readable (in my opinion) I would like to merge it.

@richox or @tustvold  do you have any concerns?

If not I plan to do over the next day or two

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs [hunk 2]
```diff
@@ -660,6 +609,127 @@ impl SortPreservingMergeStream {
...
+        }
+        self.loser_tree[0] = winner;
+        self.loser_tree_adjusted = true;
+        TreeUpdate::Complete
+    }
+}
+
+/// The result of updating the loser tree. It is the same as an Option
+/// but with specific names for easier readability
+enum TreeUpdate {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:711]: I don't really see the need for this over `Poll<Result<()>>` in particular it is confusing what its semantics are w.r.t wakers

```
if let Err(e) = self.build()? {
    return Poll::Ready(Err(e))
}
```

Is not significantly more verbose

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs [hunk 3]
```diff
@@ -660,6 +609,127 @@ impl SortPreservingMergeStream {
...
+        // Ensure all non-exhausted streams have a cursor from which
+        // rows can be pulled
+        for i in 0..num_streams {
+            match self.maybe_poll_stream(cx, i) {
+                Poll::Ready(Ok(_)) => {}
+                Poll::Ready(Err(e)) => {
+                    self.aborted = true;
+                    return TreeUpdate::Error(e);
+                }
+                Poll::Pending => return TreeUpdate::Pending,
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:635]: If this method returned Poll this could just use `ready!` or even `?`

**Review by tustvold — APPROVED** [review]
I think removing TreeUpdate would further simplify this code, but I'm happy for this to go in as is

**Pull request merged**